modyn/supervisor/internal/supervisor.py

```diff
@@ -0,0 +1,368 @@
+import json
+import logging
+import multiprocessing as mp
+import os
+import pathlib
+import threading
+import time
+from multiprocessing import Manager, Process
+from typing import Any, Optional
+
+from modyn.metadata_database.metadata_database_connection import MetadataDatabaseConnection
+from modyn.metadata_database.utils import ModelStorageStrategyConfig
+
+# pylint: disable=no-name-in-module
+from modyn.selector.internal.grpc.generated.selector_pb2 import JsonString as SelectorJsonString
+from modyn.selector.internal.grpc.generated.selector_pb2 import StrategyConfig
+from modyn.supervisor.internal.evaluation_result_writer import JsonResultWriter, TensorboardResultWriter
+from modyn.supervisor.internal.grpc_handler import GRPCHandler
+from modyn.supervisor.internal.pipeline_executor import execute_pipeline
+from modyn.supervisor.internal.utils import PipelineInfo
+from modyn.utils import is_directory_writable, model_available, trigger_available, validate_yaml
+
+logger = logging.getLogger(__name__)
+
+
+def pipeline_monitor(pipeline_process_dict: dict[int, PipelineInfo]) -> None:
+    logger.info(f"[{os.getpid()}][pipeline_monitor] start")
+    while True:
+        num_pipelines = len(pipeline_process_dict)
+        if num_pipelines > 0:
+            logger.info(f"[{os.getpid()}][pipeline_monitor] {num_pipelines} pipelines registered")
+
+        # join all the terminated procs
+        num_active_pipeline_processes = len(mp.active_children())
```

Comment:
I am a bit unsure about the use of mp.active_children here. Maybe it does exactly what it should, but for example gRPC may spawn threads etc - does this truly only list the threads/processes we spawned using multiprocessing? Maybe we should use something like `len([p_id for  for p_id, p_info in pipeline_process_dict.items() if p_info.process_handler.is_alive()])`? Not sure...

Reply:
(My main "concern" is that we have multiple supervisors due to multiprocessing of the gRPC server and I am not sure how much they interfere with each other here)

Reply:
I'll double check

Reply:
done

Reply:
Thank you!